src/utils/pushforward.jl

```diff
@@ -1,53 +1,53 @@
 """
-    Pushforward{L,G}
+    Pushforward <: AbstractLayer
 
-Differentiable pushforward of a probabilistic `layer` with an arbitrary function `post_processing`.
+Differentiable pushforward of a probabilistic optimization layer with an arbitrary function post-processing function.
 
 `Pushforward` can be used for direct regret minimization (aka learning by experience) when the post-processing returns a cost.
 
 # Fields
-- `layer::L`: anything that implements `compute_probability_distribution(layer, θ; kwargs...)`
-- `post_processing::P`: callable
+- `optimization_layer::AbstractOptimizationLayer`: probabilistic optimization layer
+- `post_processing`: callable
 
 See also: [`FixedAtomsProbabilityDistribution`](@ref).
 """
-struct Pushforward{L,P}
-    layer::L
+struct Pushforward{O<:AbstractOptimizationLayer,P} <: AbstractLayer
+    optimization_layer::O
     post_processing::P
 end
 
 function Base.show(io::IO, pushforward::Pushforward)
-    (; layer, post_processing) = pushforward
-    return print(io, "Pushforward($layer, $post_processing)")
+    (; optimization_layer, post_processing) = pushforward
+    return print(io, "Pushforward($optimization_layer, $post_processing)")
 end
 
 """
     compute_probability_distribution(pushforward, θ)
 
-Output the distribution of `pushforward.post_processing(X)`, where `X` follows the distribution defined by `pushforward.layer` applied to `θ`.
+Output the distribution of `pushforward.post_processing(X)`, where `X` follows the distribution defined by `pushforward.optimization_layer` applied to `θ`.
 
 This function is not differentiable if `pushforward.post_processing` isn't.
 
 See also: [`apply_on_atoms`](@ref).
 """
 function compute_probability_distribution(pushforward::Pushforward, θ; kwargs...)
-    (; layer, post_processing) = pushforward
-    probadist = compute_probability_distribution(layer, θ; kwargs...)
+    (; optimization_layer, post_processing) = pushforward
```

Comment:
If `PushForward` is a `LossLayer`, we might not need this method (I think it is not covered by tests).

Other question on the fly : is `the compress_distribution!` method needed, I also think it is not covered by tests.

Reply:
It is mostly used in my visualization notebooks, we can remove it. It's also dangerous because it decouples the Z samples from the proba weight, so I would prefer that

Reply:
As for the pushforward, I haven't settled on what it can be, let's open an issue to keep track